src/graph/validator/LookupValidator.cpp

```diff
@@ -54,51 +54,148 @@ Status LookupValidator::prepareFrom() {
+}
+
+Status LookupValidator::validateYieldEdge() {
+  auto yield = sentence()->yieldClause();
+  auto yieldExpr = lookupCtx_->yieldExpr;
+  for (auto col : yield->columns()) {
+    if (ExpressionUtils::hasAny(col->expr(),
+                                {Expression::Kind::kPathBuild, Expression::Kind::kVertex})) {
+      return Status::SemanticError("illegal yield clauses `%s'", col->toString().c_str());
+    }
+    if (ExpressionUtils::hasAny(col->expr(), {Expression::Kind::kEdge})) {
+      NG_RETURN_IF_ERROR(extractSchemaProp());
+    }
+    if (col->expr()->kind() == Expression::Kind::kLabelAttribute) {
+      const auto& schemaName = static_cast<LabelAttributeExpression*>(col->expr())->left()->name();
+      if (schemaName != sentence()->from()) {
+        return Status::SemanticError("Schema name error: %s", schemaName.c_str());
+      }
+    }
+    col->setExpr(ExpressionUtils::rewriteLabelAttr2EdgeProp(col->expr()));
+    NG_RETURN_IF_ERROR(invalidLabelIdentifiers(col->expr()));
+
+    auto colExpr = col->expr();
+    auto typeStatus = deduceExprType(colExpr);
+    NG_RETURN_IF_ERROR(typeStatus);
+    outputs_.emplace_back(col->name(), typeStatus.value());
+    yieldExpr->addColumn(col->clone().release());
+    NG_RETURN_IF_ERROR(deduceProps(colExpr, exprProps_));
+  }
+  return Status::OK();
+}
+
+Status LookupValidator::validateYieldTag() {
+  auto yield = sentence()->yieldClause();
+  auto yieldExpr = lookupCtx_->yieldExpr;
+  for (auto col : yield->columns()) {
+    if (ExpressionUtils::hasAny(col->expr(),
+                                {Expression::Kind::kPathBuild, Expression::Kind::kEdge})) {
+      return Status::SemanticError("illegal yield clauses `%s'", col->toString().c_str());
+    }
+    if (ExpressionUtils::hasAny(col->expr(), {Expression::Kind::kVertex})) {
+      NG_RETURN_IF_ERROR(extractSchemaProp());
+    }
+    if (col->expr()->kind() == Expression::Kind::kLabelAttribute) {
+      const auto& schemaName = static_cast<LabelAttributeExpression*>(col->expr())->left()->name();
+      if (schemaName != sentence()->from()) {
+        return Status::SemanticError("Schema name error: %s", schemaName.c_str());
+      }
+    }
+    col->setExpr(ExpressionUtils::rewriteLabelAttr2TagProp(col->expr()));
+    NG_RETURN_IF_ERROR(invalidLabelIdentifiers(col->expr()));
+
+    auto colExpr = col->expr();
+    auto typeStatus = deduceExprType(colExpr);
+    NG_RETURN_IF_ERROR(typeStatus);
+    outputs_.emplace_back(col->name(), typeStatus.value());
+    yieldExpr->addColumn(col->clone().release());
+    NG_RETURN_IF_ERROR(deduceProps(colExpr, exprProps_));
+  }
+  return Status::OK();
+}
+
+Status LookupValidator::validateYield() {
+  auto pool = qctx_->objPool();
+  auto* newCols = pool->add(new YieldColumns());
+  lookupCtx_->yieldExpr = newCols;
   if (lookupCtx_->isEdge) {
+    idxReturnCols_.emplace_back(kSrc);
+    idxReturnCols_.emplace_back(kDst);
+    idxReturnCols_.emplace_back(kType);
```

Comment:
Do we need to return EdgeType ?

Reply:
need EdgeType when  YIELD edge

Reply:
> need EdgeType when YIELD edge

I think you can get the EdgeType from graph layer. you can find edgeType by indexID. 
this will cause additional net|memory load if pushing down to the storage layer.

Reply:
we use iterator to collect edge, so the returned result needs to include edgetype,